x-pack/platform/plugins/shared/observability_ai_assistant/server/service/distributed_lock_manager/index.ts

```diff
@@ -0,0 +1,181 @@
+  constructor(private lockId: LockId, private esClient: Client, private logger: Logger) {}
+
+  /**
+   * Attempts to acquire a lock by creating a document with the given lockId.
+   * If the lock exists and is expired, it will be released and acquisition retried.
+   */
+  public async acquire(options: AcquireOptions = {}): Promise<boolean> {
+    const ttl = options.ttlMs ?? 3600 * 1000; // Default TTL: 1 hour.
+    const now = new Date();
+
+    try {
+      await this.esClient.index({
+        index: INDEX_NAME,
+        id: this.lockId,
+        body: {
+          createdAt: now.toISOString(),
+          expiresAt: new Date(now.getTime() + ttl).toISOString(),
+          meta: options.meta ?? {},
+        },
+        op_type: 'create', // Only create if the document does not already exist.
+      });
+      this.logger.debug(`Lock "${this.lockId}" acquired.`);
+      return true;
+    } catch (error: unknown) {
+      const isVersionConflictError =
+        error instanceof errors.ResponseError &&
+        error.body?.error?.type === 'version_conflict_engine_exception';
+
+      if (isVersionConflictError) {
+        const existingLock = await this.get();
+        if (!existingLock) {
+          // If the lock doesn't exist, it might have been purged because it expired and we will try acquiring it again.
+          return this.acquire(options);
+        }
+        return false;
+      }
+
+      throw error;
+    }
+  }
+
+  private isExpired(lock: LockDocument): boolean {
+    const { expiresAt } = lock;
+    return new Date(expiresAt) < new Date();
+  }
+
+  /**
+   * Releases a lock by deleting its document.
+   * If the document is not found, the error is ignored.
+   */
+  public async release(): Promise<void> {
+    await this.esClient.delete({ index: INDEX_NAME, id: this.lockId }, { ignore: [404] });
+    this.logger.debug(`Lock "${this.lockId}" released.`);
+  }
+
+  /**
+   * Retrieves the lock document for a given lockId.
+   * If the lock is expired, it will be released.
+   */
+  public async get(): Promise<LockDocument | undefined> {
+    const res = await this.esClient.get<LockDocument>(
+      { index: INDEX_NAME, id: this.lockId },
+      { ignore: [404] }
+    );
+
+    if (!res?._source) {
+      return;
+    }
+
+    // Check if the lock is expired.
+    if (this.isExpired(res?._source)) {
+      await this.release();
+      return;
+    }
+
+    return res._source;
+  }
+
+  public async waitForLock(options: AcquireOptions & pRetry.Options = {}): Promise<boolean> {
+    const { meta, ttlMs, ...pRetryOptions } = options;
+
+    return pRetry(async () => {
+      const acquired = await this.acquire({ meta, ttlMs });
+      if (!acquired) {
+        this.logger.debug(`Lock "${this.lockId}" not available yet.`);
+        throw new Error(`Lock "${this.lockId}" not available yet`);
+      }
+      return acquired;
+    }, pRetryOptions ?? { forever: true, maxTimeout: 10_000 });
+  }
+}
+
+export async function withLock<T>(
```

Comment:
Could this live in the LockManager class, and consumer would use it as `await taskManager.withLock(someFn, { opts })` ?

Reply:
I'm ++ on making this a method exposed by a client and not a separate function that is imported and requires all the dependencies, it makes it harder to use IMHO

Reply:
Fundamentally I think this should be the main way to interact, instead of it being exposed as a utility function (at least, I don't understand why the other methods are needed)

Reply:
I've exposed `withLock` as a service method:

```ts
const lockManagerService = new LockManagerService(coreSetup, logger);
lockManagerService.withLock('my_lock', () => {
 // perform operation
});
```

WDYT about that? The stand alone methods (`acquire`, `release` etc) are not exposed on the service and currently only used directly for testing purposes.

Reply:
sgtm!